pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -670,16 +687,37 @@ impl LayerInner {
                 Ok(ResidentOrWantedEvicted::Resident(res))
             };
 
-            let locked = self.inner.get_or_init(download).await?;
+            let (weak, _permit) = {
+                // should we be able to give the permit to the `get_or_init`? would make sense.
+                drop(permit.take());
+                let mut locked = self.inner.get_or_init(download).await?;
 
-            if let Some(strong) = Self::get_or_apply_evictedness(Some(locked), &self.wanted_evicted)
-            {
-                return Ok(strong);
-            }
+                if let Some(strong) = locked.get_and_upgrade() {
+                    self.wanted_evicted.store(false, Ordering::Relaxed);
+
+                    // error out any `evict_and_wait`
+                    drop(self.status.send(Status::Downloaded));
+
+                    return Ok(strong);
+                } else {
+                    // path to here: the evict_blocking is stuck on spawn_blocking queue.
```

Comment:
This makes sense with careful reading.  It will be interesting to see how common the case is where evictions and downloads are racing: if it is rare, we might be able to simplify this overall control flow by having downloads wait for evictions to complete before starting, rather than accommodating the race.

Reply:
Well, that would be a bigger design change at least, and how it has been on staging for the past days. There has been at least one looping around of this loop in staging, but that is of course not much :)

Theoretically we could start the deinit from the `DownloadedLayer::drop` *however* there's a deadlock risk there. I think avoiding the deadlock by doing a spawn_blocking is of importance because those deadlocks might not be hit immediatedly in local testing or even regress suite (I don't claim this code is deadlock free right now, but I'm quite sure it is).

Note that in this branch we completely cancel the eviction and will not incur a download penalty.